File: backend/app/routers/gmail.py
"""
Gmail OAuth routes: connect, callback, status, disconnect.
"""

from datetime import datetime, timedelta, timezone

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import RedirectResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..database import get_db
from ..models import GmailToken
from ..schemas import GmailStatusResponse
from ..services.gmail_service import exchange_code, get_auth_url, get_user_email

router = APIRouter(prefix="/gmail", tags=["gmail"])


async def _get_token(db: AsyncSession) -> GmailToken | None:
    """Get the single stored Gmail token (single-user app)."""
    result = await db.execute(select(GmailToken).order_by(GmailToken.id.desc()).limit(1))
    return result.scalars().first()


@router.get("/connect")
async def gmail_connect():
    """Redirect user to Google OAuth consent screen."""
    return RedirectResponse(url=get_auth_url())


@router.get("/callback")
async def gmail_callback(code: str | None = None, error: str | None = None, db: AsyncSession = Depends(get_db)):
    """Handle Google OAuth callback."""
    if error:
        return RedirectResponse(url="http://localhost:3000?gmail_error=" + error)

    if not code:
        raise HTTPException(status_code=400, detail="Missing authorization code")

    try:
        token_data = await exchange_code(code)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Token exchange failed: {e}")

    access_token = token_data["access_token"]
    refresh_token = token_data.get("refresh_token", "")
    expires_in = token_data.get("expires_in", 3600)
    token_expiry = datetime.now(timezone.utc) + timedelta(seconds=expires_in)

    # Get user email
    email = ""
    try:
        email = await get_user_email(access_token)
    except Exception:
        pass

    # Upsert: delete old tokens, insert new one (single-user)
    old_tokens = await db.execute(select(GmailToken))
    for old in old_tokens.scalars().all():
        await db.delete(old)

    gmail_token = GmailToken(
        access_token=access_token,
        refresh_token=refresh_token,
        token_expiry=token_expiry,
        email=email,
    )
    db.add(gmail_token)
    await db.commit()

    # Redirect back to frontend
    return RedirectResponse(url="http://localhost:3000?gmail_connected=true")


@router.get("/status", response_model=GmailStatusResponse)
async def gmail_status(db: AsyncSession = Depends(get_db)):
    """Check if Gmail is connected."""
    token = await _get_token(db)
    if not token:
        return GmailStatusResponse(connected=False)
    return GmailStatusResponse(connected=True, email=token.email)


@router.post("/disconnect", status_code=204)
async def gmail_disconnect(db: AsyncSession = Depends(get_db)):
    """Remove stored Gmail tokens."""
    tokens = await db.execute(select(GmailToken))
    for token in tokens.scalars().all():
        await db.delete(token)
    await db.commit()

File: server/src/routes/gmail.js
// Gmail OAuth routes: connect (redirect to Google), handle callback,
// check connection status, and disconnect.
// All routes require authentication — tokens are stored per user.

var express = require("express");
var jwt = require("jsonwebtoken");
var { getPool } = require("../database");
var config = require("../config");
var gmailService = require("../services/gmailService");
var requireAuth = require("../middleware/auth");

var router = express.Router();

// Redirect the user to Google's OAuth consent screen.
// Accepts the JWT token as a query parameter since browser redirects can't set headers.
router.get("/connect", function (req, res) {
  // token comes from the ?token= query param set by the frontend
  var token = req.query.token;

  if (!token) {
    res.status(401).json({ detail: "Missing token. Please log in first." });
    return;
  }

  // Verify the token is valid before starting the OAuth flow
  try {
    jwt.verify(token, config.jwtSecret);
  } catch (e) {
    res.status(401).json({ detail: "Invalid or expired token. Please log in again." });
    return;
  }

  // authUrl is the Google consent page URL
  var authUrl = gmailService.getAuthUrl();

  // Pass the JWT as OAuth state so the callback can identify the user
  res.redirect(authUrl + "&state=" + encodeURIComponent(token));
});

// Handle the callback from Google after the user grants permission.
// Stores the tokens in the database linked to the user, then redirects to the frontend.
router.get("/callback", async function (req, res, next) {
  try {
    var db = getPool();

    // error and code are query parameters Google sends back
    var error = req.query.error;
    var code = req.query.code;
    // state contains the JWT token we passed in /connect
    var state = req.query.state;

    if (error) {
      res.redirect("http://localhost:3000?gmail_error=" + error);
      return;
    }

    if (!code) {
      res.status(400).json({ detail: "Missing authorization code" });
      return;
    }

    // Figure out which user started this OAuth flow by decoding the state token
    var userId = null;
    if (state) {
      try {
        var decoded = jwt.verify(state, config.jwtSecret);
        userId = decoded.userId;
      } catch (e) {
        res.status(401).json({ detail: "Invalid or expired session. Please log in and try again." });
        return;
      }
    }

    if (!userId) {
      res.status(401).json({ detail: "Could not identify user. Please log in and try again." });
      return;
    }

    // tokenData contains access_token, refresh_token, expires_in, etc.
    var tokenData;
    try {
      tokenData = await gmailService.exchangeCode(code);
    } catch (e) {
      res.status(400).json({ detail: "Token exchange failed: " + e.message });
      return;
    }

    // Pull out the token fields we need to store
    var accessToken = tokenData.access_token;
    var refreshToken = tokenData.refresh_token || "";
    var expiresIn = tokenData.expires_in || 3600;

    // tokenExpiry is the timestamp when this access token will stop working
    var tokenExpiry = new Date(Date.now() + expiresIn * 1000);

    // Try to get the user's Gmail address
    var email = "";
    try {
      email = await gmailService.getUserEmail(accessToken);
    } catch (e) {
      // not critical — we'll just store an empty email
    }

    // Delete any existing tokens for THIS user (each user gets one Gmail connection)
    await db.query("DELETE FROM gmail_tokens WHERE user_id = $1", [userId]);

    // Insert the new token linked to this user
    await db.query(
      `INSERT INTO gmail_tokens (user_id, access_token, refresh_token, token_expiry, email)
       VALUES ($1, $2, $3, $4, $5)`,
      [userId, accessToken, refreshToken, tokenExpiry, email]
    );

    // Send the user back to the frontend with a success flag
    res.redirect("http://localhost:3000?gmail_connected=true");
  } catch (err) {
    next(err);
  }
});

// Check whether Gmail is connected for the logged-in user
router.get("/status", requireAuth, async function (req, res, next) {
  try {
    var db = getPool();
    var userId = req.user.id;

    var result = await db.query(
      "SELECT * FROM gmail_tokens WHERE user_id = $1 ORDER BY id DESC LIMIT 1",
      [userId]
    );

    // token is the stored Gmail token for this user, or undefined if not connected
    var token = result.rows[0];

    if (!token) {
      res.json({ connected: false, email: null });
      return;
    }

    res.json({ connected: true, email: token.email });
  } catch (err) {
    next(err);
  }
});

// Disconnect Gmail by removing the logged-in user's stored tokens
router.post("/disconnect", requireAuth, async function (req, res, next) {
  try {
    var db = getPool();
    var userId = req.user.id;

    await db.query("DELETE FROM gmail_tokens WHERE user_id = $1", [userId]);
    res.status(204).send();
  } catch (err) {
    next(err);
  }
});

module.exports = router;
